extensions/consensus/consensus.go

```diff
@@ -0,0 +1,240 @@
+	_, have := Hardforks[name]
+	return have
+}
+
+// Hardfork specifies the fundamental changes affected by a named hardfork. If
+// a field is nil or the zero value for the type, that particular change is not
+// part of the hardfork's definition. When the hardfork is registered, only the
+// defined changes are applied in the relevant parts of kwild.
+type Hardfork struct {
+	// Name is the hard fork's code name.
+	Name string
+	// NOTE: Activation height is specified by genesis.json (or other dynamic
+	// methods like signaling / voting). This struct defines the changes.
+
+	// Optional consensus logic overrides below.  All of these will require
+	// dynamic kwil-db support, such as registering new encoders and transaction
+	// payloads, submitting consensus parameter updates to the consensus engine,
+	// applying state adjustments at activation height in ABCI / TxApp, etc.
+
+	// TxPayloads specifies new transaction payload to recognize at activation.
+	// To modify an existing payload in a backward incompatible way, instead
+	// create a new version of the payload such as PayloadTypeExecuteActionV2.
+	// TODO: consider payload removal/invalidation and replace.
+	TxPayloads []Payload // Type() and tx app Route() implementation
+
+	// ResolutionUpdates are coordinated change to the resolutions extension.
+	// They may be added, modified (redefined), or removed at activation. Any
+	// have the potential to break consensus, and should be done with a Hardfork.
+	ResolutionUpdates []*ResolutionMod
+
+	// Encoders are new encoder types to register *at activation height* for
+	// core/types/serialize.Encode/Decode e.g. Borsch instead of RLP.
+	// EncodingTypeCustom offset should be used as the first possible external
+	// codec's type to leave space for more kwild canonical codecs in the
+	// future. Choose an encoding type that does not collide with other codecs.
+	Encoders []*serialize.Codec // unimplemented!
+
+	// ParamsUpdates are updates to the consensus engine parameters that should
+	// go into affect at after the activation height (returned to the consensus
+	// engine when finalizing the block at this height).
+	ParamsUpdates *ParamUpdates
+
+	// StateMod is triggered at activation. It can do anything, one time. For
+	// instance, arbitrary change to application state via the Engine or more
+	// directly to the DB may be made at the end of block at the activation
+	// height. This is to be called inside the outer transaction of activation
+	// block, so changes to date state are captured in the normal apphash diff.
+	// This is a reasonable capability for a hardfork to make state changes
+	// outside of transaction execution, but most such changes can probably be
+	// achieved through the resolution system and voting. Doing it in a hardfork
+	// would be needed if there are other changes (either baked in to kwild or
+	// in the above fields) that should be done in concert with the fork.
+	StateMod func(context.Context, *common.App) error
```

Comment:
To give a concrete instance of this, there was an etherum hard fork ["tangerine whistle"](https://blog.ethereum.org/2016/10/18/faq-upcoming-ethereum-hard-fork) that included in its' changes clearing of empty accounts to eliminate state bloat caused by an attack.  Rule changes were complementary to the state modification.  TheDAO hard fork also included state modifications to rectify issues.